tests/tenancy_test.go

```diff
@@ -0,0 +1,150 @@
+	operators := operatorBuilder(2, true, true)
+	for _, operator := range operators {
+		err := operator.InstallSparkOperator(true)
+		assert.NilError(t, err)
+	}
+
+	verifyComponents(t, operators)
+	verifyWorkloads(t, operators)
+
+	// verify CRDs are present after one of the operators is deleted
+	operators[0].CleanUp()
+	assert.Assert(t, crdsInstalled(t), "CRDs are not present!")
+
+	// check that CRDs are deleted after no operator instances left
+	operators[1].CleanUp()
+	assert.Assert(t, !crdsInstalled(t), "CRDs are not deleted!")
+}
+
+func TestTenancyTwoOperatorsSingleNamespace(t *testing.T) {
+	operators := operatorBuilder(2, false, true)
+	for _, operator := range operators {
+		err := operator.InstallSparkOperator(false)
+		assert.NilError(t, err)
+		defer operator.CleanUp()
+	}
+
+	verifyComponents(t, operators)
+	verifyWorkloads(t, operators)
+}
+
+func TestTenancyTwoOperatorsSameNameDifferentNamespaces(t *testing.T) {
+	operators := operatorBuilder(2, true, false)
+	for _, operator := range operators {
+		err := operator.InstallSparkOperator(true)
+		assert.NilError(t, err)
+	}
+
+	verifyComponents(t, operators)
+	verifyWorkloads(t, operators)
+
+}
+
+func verifyComponents(t *testing.T, operators []*utils.SparkOperatorInstallation) {
+	serviceAccounts := []string{"spark-operator-service-account", "spark-service-account"}
+	services := []string{"spark-webhook", "spark-operator-metrics"}
+	roles := []string{"spark-driver-role", "%s-spark-role"}
+
+	for _, operator := range operators {
+		for _, service := range services {
+			serviceName := fmt.Sprint(operator.InstanceName, "-", service)
+			log.Infof("Checking Service \"%s\" is created in namespace \"%s\" for \"%s\"", serviceName,
+				operator.Namespace, operator.InstanceName)
+			result, err := operator.K8sClients.CoreV1().Services(operator.Namespace).Get(fmt.Sprint(serviceName), v1.GetOptions{})
+			assert.NilError(t, err)
+			assert.Equal(t, result.Labels["kudo.dev/instance"], operator.InstanceName)
+		}
+
+		for _, sa := range serviceAccounts {
+			serviceAccount := fmt.Sprint(operator.InstanceName, "-", sa)
+			log.Infof("Checking ServiceAccount \"%s\" is created in namespace \"%s\" for \"%s\"", serviceAccount,
+				operator.Namespace, operator.InstanceName)
+			result, err := operator.K8sClients.CoreV1().ServiceAccounts(operator.Namespace).Get(serviceAccount, v1.GetOptions{})
+			assert.NilError(t, err)
+			assert.Equal(t, result.Labels["kudo.dev/instance"], operator.InstanceName)
+		}
+
+		for _, role := range roles {
+			if strings.Contains(role, "%s") {
+				role = fmt.Sprintf(role, operator.InstanceName)
+			}
+			log.Infof("Checking Role \"%s\" is created in namespace \"%s\" for \"%s\"",
+				role, operator.Namespace, operator.InstanceName)
+			result, err := operator.K8sClients.RbacV1().Roles(operator.Namespace).Get(role, v1.GetOptions{})
+			assert.NilError(t, err)
+			instanceLabel, present := result.Labels["kudo.dev/instance"]
+			if present {
```

Comment:
Why if the label is not present here, the situation is ignored?

Reply:
yes, because in this particular case not all the `Role` resources contain this label that we are verifying, therefore we can't explicitly check to which instance it belongs to. But we are still checking the role exists it the operator's namespace, so this test is valid.

Reply:
ok, got it